pkg/log/formatter.go

```diff
@@ -0,0 +1,84 @@
+package log
+
+import (
+	"fmt"
+	"github.com/applike/gosoline/pkg/encoding/json"
+	"github.com/fatih/color"
+	"sort"
+	"strings"
+)
+
+type Formatter func(timestamp string, level int, format string, args []interface{}, err error, data Data) ([]byte, error)
+
+var formatters = map[string]Formatter{
+	"console": FormatterConsole,
+	"json":    FormatterJson,
+}
+
+func FormatterConsole(timestamp string, level int, format string, args []interface{}, err error, data Data) ([]byte, error) {
+	fieldString := getFieldsAsString(data.Fields)
+	contextString := getFieldsAsString(data.ContextFields)
+
+	levelStr := fmt.Sprintf("%-7s", LevelName(level))
+	channel := fmt.Sprintf("%-7s", data.Channel)
+	msg := fmt.Sprintf(format, args...)
+
+	if err != nil {
+		msg = color.RedString(err.Error())
+	}
+
+	output := fmt.Sprintf("%s %s %s %-50s %s %s",
+		color.YellowString(timestamp),
+		color.GreenString(channel),
+		color.GreenString(levelStr),
+		msg,
+		color.GreenString(contextString),
+		color.BlueString(fieldString),
+	)
+
+	output = strings.TrimSpace(output)
+	serialized := []byte(output)
+
+	return append(serialized, '\n'), nil
+}
+
+func FormatterJson(timestamp string, level int, format string, args []interface{}, err error, data Data) ([]byte, error) {
+	msg := fmt.Sprintf(format, args...)
+	jsn := make(map[string]interface{}, 8)
+
+	if err != nil {
+		jsn["err"] = err.Error()
+	}
+
+	jsn["channel"] = data.Channel
```

Comment:
Wouldn't it be nicer if we would be using a go struct with the correct annotations (e.g., omit_empty for errors) and nice types?

Reply:
Is it more performant using a map?  
Struct would involve costly reflection usage with every log message? :thinking:

Reply:
Not sure what is going on in go here... but for me a struct is actually twice as fast as a map (3us instead of 6us/iteration). I am just using 

```go
type JsonStruct struct {
	Error     string                 `json:"err,omitempty"`
	Channel   string                 `json:"channel"`
	Level     int                    `json:"level"`
	LevelName string                 `json:"level_name"`
	Timestamp string                 `json:"timestamp"`
	Message   string                 `json:"message"`
	Fields    map[string]interface{} `json:"fields"`
	Context   map[string]interface{} `json:"context"`
}
```

so there isn't going anything fancy on here... maybe it can allocate the struct on the stack or something like that? Or reflection isn't as expensive here?

Reply:
okay, switched to struct